flytekit/annotated/reference_entity.py

```diff
@@ -0,0 +1,200 @@
+from typing import Any, Dict, Tuple, Type, Union
+
+from flytekit.annotated.context_manager import BranchEvalMode, ExecutionState, FlyteContext
+from flytekit.annotated.interface import Interface, transform_interface_to_typed_interface
+from flytekit.annotated.node import create_and_link_node
+from flytekit.annotated.promise import Promise, VoidPromise, create_task_output, translate_inputs_to_literals
+from flytekit.annotated.type_engine import TypeEngine
+from flytekit.common.exceptions import user as _user_exceptions
+from flytekit.loggers import logger
+from flytekit.models import dynamic_job as _dynamic_job
+from flytekit.models import interface as _interface_models
+from flytekit.models import literals as _literal_models
+from flytekit.models.core import identifier as _identifier_model
+
+
+class Reference(object):
+    def __init__(
+        self, resource_type: int, project: str, domain: str, name: str, version: str, *args, **kwargs,
+    ):
+        self._id = _identifier_model.Identifier(resource_type, project, domain, name, version)
+
+    @property
+    def id(self) -> _identifier_model.Identifier:
+        return self._id
+
+
+class TaskReference(Reference):
```

Comment:
these could all be simple derived dataclasses

Reply:
This design is a bit weird and not so consistent. I guess it is due to we use both class hierarchy and  property to differentiate types, so there are cases `Reference(_identifier_model.ResourceType.TASK, "foo", "bar", "foo.bar", "v") != TaskReference("foo", "bar", "foo.bar", "v")`, for example when looking up a task plugin. I haven't looked into details of this PR, but may it solves it?

Reply:
We decided to nix the looking up of task plugins and just go with an explicit `@reference_task` decorator.  

The inequality you mention is a bit unfortunate yeah, do you have any recommendations?

Thanks.

Reply:
I think what @honnix wants to say probably is that the base type should not be instantiable on its own?

Reply:
Yeah that could be a solution if we make Reference abstract. But this is not a big deal and we don't need to block on it.